Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/rendering/

cdumez opened the pull request:
#### 52499fd4a1ee333480ec4b673df2c6a2d7ad31de
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/rendering/
<a href="https://bugs.webkit.org/show_bug.cgi?id=283274">https://bugs.webkit.org/show_bug.cgi?id=283274</a>

Reviewed by Darin Adler.

* Source/WebCore/html/HTMLFrameSetElement.h:
* Source/WebCore/rendering/AutoTableLayout.cpp:
(WebCore::AutoTableLayout::insertSpanCell):
* Source/WebCore/rendering/BreakLines.cpp:
* Source/WebCore/rendering/BreakLines.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::DefiniteSizeStrategy::maximizeTracks):
* Source/WebCore/rendering/RenderDetailsMarker.cpp:
(WebCore::createPath):
(WebCore::createDownArrowPath):
(WebCore::createUpArrowPath):
(WebCore::createLeftArrowPath):
(WebCore::createRightArrowPath):
* Source/WebCore/rendering/RenderFrameSet.cpp:
(WebCore::RenderFrameSet::layOutAxis):
* Source/WebCore/rendering/RenderFrameSet.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::patternForTouchAction):
* Source/WebCore/rendering/RenderQuote.cpp:
(WebCore::checkNumberOfDistinctQuoteCharacters):
(WebCore::strcmpSpan):
(WebCore::subtagCompare):
(WebCore::quoteTableLanguageComparisonFunction):
(WebCore::quotesForLanguage):
(WebCore::stringForQuoteCharacter):
* Source/WebCore/rendering/RenderTableCell.cpp:

Canonical link: <a href="https://commits.webkit.org/286774@main">https://commits.webkit.org/286774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2d59117f87c6215131d15637872090f3d17c291

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77046 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56081 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29961 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81596 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28325 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79163 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65229 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4377 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/18443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80113 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50325 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66136 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/40687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47727 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23634 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26651 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70231 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68843 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23962 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83030 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76324 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4426 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2971 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68655 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66109 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67908 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11891 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9973 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98577 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11920 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4372 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21569 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/4392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/7827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6151 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/AutoTableLayout.cpp:
```diff
@@ -482,8 +480,8 @@ void AutoTableLayout::insertSpanCell(RenderTableCell *cell)
-        pos++;
-    memmove(m_spanCells.data()+pos+1, m_spanCells.data()+pos, (size-pos-1)*sizeof(RenderTableCell *));
+        ++pos;
+    memmove(m_spanCells.mutableSpan().subspan(pos + 1).data(), m_spanCells.subspan(pos).data(), (size - pos - 1) * sizeof(RenderTableCell*));
```
I think `(size - (pos + 1))` would be more logical. And there must be a safer way to write memory moves where we don’t have to do the math. Maybe we need to make a `moveWithinSpan` function alongside `memcpySpan`.

darinadler commented on Source/WebCore/rendering/RenderQuote.cpp:
```diff
@@ -99,28 +98,36 @@ struct SubtagComparison {
+    if (!result && s1.size() != s2.size())
+        result = (s1.size() > s2.size()) ? 1 : -1;
+    return result;
+}
```
I suppose we can move this into a header later, when we find somewhere else we need it.

darinadler commented on Source/WebCore/rendering/RenderQuote.cpp:
```diff
@@ -99,28 +98,36 @@ struct SubtagComparison {
-    if (auto* hyphenPointer = strchr(key, '-')) {
-        result.keyLength = hyphenPointer - key;
+    if (auto* hyphenPointer = memchr(key.data(), '-', key.size())) {
+        result.keyLength = reinterpret_cast<uintptr_t>(hyphenPointer) - reinterpret_cast<uintptr_t>(key.data());
```
```suggestion
        result.keyLength = static_cast<LChar*>(hyphenPointer) - key.data();
```

cdumez replied:
Ok

darinadler commented on Source/WebCore/rendering/RenderQuote.cpp:
```diff
@@ -99,28 +98,36 @@ struct SubtagComparison {
-    if (auto* hyphenPointer = strchr(range, '-')) {
-        result.rangeLength = hyphenPointer - range;
+    if (auto* hyphenPointer = memchr(range.data(), '-', range.size())) {
+        result.rangeLength = reinterpret_cast<uintptr_t>(hyphenPointer) - reinterpret_cast<uintptr_t>(range.data());
```
```suggestion
        result.rangeLength = static_cast<LChar*>(hyphenPointer) - range.data();
```

cdumez replied:
Ok

darinadler commented on Source/WebCore/rendering/RenderQuote.cpp:
```diff
@@ -203,160 +211,161 @@ static int quoteTableLanguageComparisonFunction(const void* a, const void* b)
-        { "zh"_s,         1, 0x201c, 0x201d, 0x2018, 0x2019 },
-        { "zh-hant"_s,    0, 0x300c, 0x300d, 0x300e, 0x300f },
-        { "zu"_s,         0, 0x201c, 0x201d, 0x2018, 0x2019 },
+    static const std::array quoteTable {
```
```suggestion
    static constexpr std::array quoteTable {
```

cdumez replied:
This doesn't build at the moment:
```
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource397.cpp:6:
/Volumes/Work/WebKit/OpenSource/Source/WebCore/rendering/RenderQuote.cpp:213:33: error: constexpr variable 'quoteTable' must be initialized by a constant expression
  213 |     static constexpr std::array quoteTable {
  214 |         QuotesForLanguage { "af"_span,         0, 0x201c, 0x201d, 0x2018, 0x2019 },
  215 |         QuotesForLanguage { "agq"_span,        0, 0x201e, 0x201d, 0x201a, 0x2019 },
  216 |         QuotesForLanguage { "ak"_span,         0, 0x201c, 0x201d, 0x2018, 0x2019 },
  217 |         QuotesForLanguage { "am"_span,         0, 0x00ab, 0x00bb, 0x2039, 0x203a },
  218 |         QuotesForLanguage { "ar"_span,         0, 0x201d, 0x201c, 0x2019, 0x2018 },
  219 |         QuotesForLanguage { "asa"_span,        0, 0x201c, 0x201d, 0x2018, 0x2019 },
  220 |         QuotesForLanguage { "az-cyrl"_span,    0, 0x00ab, 0x00bb, 0x2039, 0x203a },
  221 |         QuotesForLanguage { "bas"_span,        0, 0x00ab, 0x00bb, 0x201e, 0x201c },
  222 |         QuotesForLanguage { "bem"_span,        0, 0x201c, 0x201d, 0x2018, 0x2019 },
  223 |         QuotesForLanguage { "bez"_span,        0, 0x201c, 0x201d, 0x2018, 0x2019 },
  224 |         QuotesForLanguage { "bg"_span,         0, 0x201e, 0x201c, 0x201a, 0x2018 },
  225 |         QuotesForLanguage { "bm"_span,         0, 0x00ab, 0x00bb, 0x201c, 0x201d },
  226 |         QuotesForLanguage { "bn"_span,         0, 0x201c, 0x201d, 0x2018, 0x2019 },
  227 |         QuotesForLanguage { "br"_span,         0, 0x00ab, 0x00bb, 0x2039, 0x203a },
  228 |         QuotesForLanguage { "brx"_span,        0, 0x201c, 0x201d, 0x2018, 0x2019 },
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource397.cpp:1:
In file included from /Volumes/Work/WebKit/OpenSource/Source/WebCore/WebCorePrefix.h:164:
In file included from /Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/FastMalloc.h:26:
/Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/StdLibExtras.h:782:25: note: reinterpret_cast is not allowed in a constant expression
  782 |     return ReturnType { reinterpret_cast<T*>(const_cast<std::remove_const_t<U>*>(span.data())), span.size_bytes() / sizeof(T) };
/Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/StdLibExtras.h:917:95: note: in call to 'spanReinterpretCast({&"af"[0], 2})'
  917 |     template<ByteType U> static constexpr auto cast(std::span<const T, Extent> span) { return spanReinterpretCast<const U>(span); }
/Volumes/Work/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/StdLibExtras.h:922:12: note: in call to 'cast({&"af"[0], 2})'
```

cdumez replied:
```
constexpr std::span<const LChar> operator""_span(const char* characters, size_t n)
{
    auto span = byteCast<LChar>(unsafeMakeSpan(characters, n));
```

`byteCast<LChar>()` ends up calling `spanReinterpretCast<>()` which calls `reinterpret_cast<>()` which is not not allowed with constexpr.

cdumez replied:
I'll try and fix this is in a follow-up. It should be able to make this truly `constexpr`.

cdumez replied:
Actually, std::bit_cast between pointer types is not constexpr so it doesn't seem trivial to fix.

darinadler commented on Source/WebCore/rendering/RenderQuote.cpp:
```diff
@@ -99,28 +98,36 @@ struct SubtagComparison {
 };
 
-static SubtagComparison subtagCompare(const char* key, const char* range)
+static int strcmpSpan(std::span<const LChar> s1, std::span<const LChar> s2)
```
I might name this `cmpSpan` rather than `strcmpSpan`, since `str` is the standard library’s name for "null-terminated array of `char`".

cdumez replied:
We already have a `equalSpans()` which wraps memcmp() for spans. However, it is only useful to compare spans of the same size. memcmp() has to specify the number of bytes we want to compare.

Here, we're really looking for a strcmp replacement which works even if the 2 strings have different sizes and will not simply compare the first N bytes. I understand though that using the term "strcmp" in the function name is a bit confusing since:
- it doesn't call strcmp() internally
- It doesn't deal with null-terminated strings.

Not quite sure how to communicate that this is basically the behavior of strcmp but with spans instead of null-terminated strings.

cdumez replied:
I think for now I'll use this unless someone has a better proposal:
```
template<typename T>
int compareSpans(std::span<const T> a, std::span<const T> b)
{
    static_assert(std::has_unique_object_representations_v<T>);
    int result = memcmp(a.data(), b.data(), std::min(a.size(), b.size()));
    if (!result && a.size() != b.size())
        result = (a.size() > b.size()) ? 1 : -1;
    return result;
}
```

darinadler replied:
That sounds good. I was thinking that “cmp” was the nickname for the -1/0/1 result. But I think “compare” is also fine. I should also note that the modern C++ way to do this is sort of thing is to use `std::strong_ordering` instead.

darinadler replied:
This seems like the correct logic for byte spans of different lengths, so I don’t think there’s any need for the word “string”.

darinadler commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -119,16 +117,17 @@ class BreakLines {
     public:
         static constexpr UChar firstCharacter = '!';
         static constexpr UChar lastCharacter = 0xFF;
+        static constexpr unsigned rowCount = lastCharacter - firstCharacter + 1;
+        static constexpr unsigned columnCount = (lastCharacter - firstCharacter) / 8 + 1;
```
What’s the rationale for making these public?

cdumez replied:
In Breaklines.cpp, the `static const std::array<std::array<uint8_t, columnCount>, rowCount> breakTable;` array is defined. However, they had hard-coded the columnCount and rowCount. I wanted to re-use the same variables as in the header (where the array is declared).

darinadler reviewed (approved):
I reviewed the rest of this, but some kind of GitHub bug prevented me from reviewing `BreakLines.cpp` changes.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebCore/rendering/
https://bugs.webkit.org/show_bug.cgi?id=283274

Reviewed by Darin Adler.

* Source/WebCore/html/HTMLFrameSetElement.h:
* Source/WebCore/rendering/AutoTableLayout.cpp:
(WebCore::AutoTableLayout::insertSpanCell):
* Source/WebCore/rendering/BreakLines.cpp:
* Source/WebCore/rendering/BreakLines.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::DefiniteSizeStrategy::maximizeTracks):
* Source/WebCore/rendering/RenderDetailsMarker.cpp:
(WebCore::createPath):
(WebCore::createDownArrowPath):
(WebCore::createUpArrowPath):
(WebCore::createLeftArrowPath):
(WebCore::createRightArrowPath):
* Source/WebCore/rendering/RenderFrameSet.cpp:
(WebCore::RenderFrameSet::layOutAxis):
* Source/WebCore/rendering/RenderFrameSet.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::patternForTouchAction):
* Source/WebCore/rendering/RenderQuote.cpp:
(WebCore::checkNumberOfDistinctQuoteCharacters):
(WebCore::strcmpSpan):
(WebCore::subtagCompare):
(WebCore::quoteTableLanguageComparisonFunction):
(WebCore::quotesForLanguage):
(WebCore::stringForQuoteCharacter):
* Source/WebCore/rendering/RenderTableCell.cpp:

Canonical link: https://commits.webkit.org/286774@main
